pkg/controllers/management/cluster/clustercontroller.go

```diff
@@ -109,11 +123,77 @@ func (c *controller) capsSync(key string, cluster *v3.Cluster) (runtime.Object,
 		if _, err := c.clusterClient.Update(toUpdateCluster); err != nil {
 			return nil, err
 		}
+		return toUpdateCluster, nil
 	}
 
 	return nil, nil
 }
 
+// overrideCapabilities masks the given Capabilities struct with values extracted from annotations prefixed with
+// "capabilities.cattle.io/"
+func overrideCapabilities(annotations map[string]string, oldCapabilities v3.Capabilities, schema *normantypes.Schema) (v3.Capabilities, error) {
+	capabilities := v3.Capabilities{}
+
+	capabilitiesMap, err := convert.EncodeToMap(oldCapabilities)
+	if err != nil {
+		return capabilities, err
+	}
+
+	var isUpdate bool
+	for annoKey, annoValue := range annotations {
+		if strings.HasPrefix(annoKey, capabilitiesAnnotation) {
+			capability := strings.TrimPrefix(annoKey, capabilitiesAnnotation)
+			val, err := parseResourceInterface(capability, annoValue, schema)
+			if err != nil {
+				return capabilities, err
+			}
+			capabilitiesMap[capability] = val
+			isUpdate = true
+		}
+
+	}
+
+	if isUpdate {
+		if err := convert.ToObj(capabilitiesMap, &capabilities); err != nil {
+			return capabilities, err
+		}
+		return capabilities, nil
+	}
+
+	return oldCapabilities, nil
+}
+
+// parseResourceInterface converts a capability annotation to the appropriate type based on given schema.
+// Json format is assumed if type is not bool, string, or integer.
+func parseResourceInterface(key string, annoValue string, schema *normantypes.Schema) (interface{}, error) {
```

Comment:
we didn't want to force the user to json encode simple values

Reply:
Yeah, fundamentally this is ok